### PR TITLE
Coinbase New API key Support

### PR DESCRIPTION
**Pull request by vijithreddy**
## Checklist

- [ ] The PR modified the frontend, and updated the [user guide](https://github.com/rotki/rotki/blob/develop/docs/usage_guide.rst) to reflect the changes.

### CODE REVIEW

File: rotkehlchen/exchanges/coinbase.py [hunk 1]
```diff
@@ -58,10 +61,8 @@
 logger = logging.getLogger(__name__)
 log = RotkehlchenLogsAdapter(logger)
 
-
 CB_EVENTS_PREFIX = 'CBE_'
-CB_VERSION = '2019-08-25'  # the latest api version we know rotki works fine for: https://docs.cloud.coinbase.com/sign-in-with-coinbase/docs/versioning  # noqa: E501
-
+CB_VERSION = '2019-08-25'
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/coinbase.py:65]: Why did you remove the comment? Is it in some way wrong?

File: rotkehlchen/exchanges/coinbase.py [hunk 2]
```diff
@@ -160,21 +161,117 @@ def __init__(
...
             database=database,
         )
+        self.is_legacy_api_key=self.determine_key_type(api_key)
         self.apiversion = 'v2'
         self.base_uri = 'https://api.coinbase.com'
         self.msg_aggregator = msg_aggregator
-        self.session.headers.update({'CB-ACCESS-KEY': self.api_key})
-        self.session.headers.update({'CB-VERSION': CB_VERSION})
+        self.host = 'api.coinbase.com'
+
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/coinbase.py:169]: Never two empty lines

```suggestion
```

File: rotkehlchen/exchanges/coinbase.py [hunk 3]
```diff
@@ -160,21 +161,117 @@ def __init__(
...
+        
+        if self.is_legacy_api_key:
+            self.session.headers.update({'CB-ACCESS-KEY': self.api_key})
+            self.session.headers.update({'CB-VERSION': CB_VERSION})
+
+    def determine_key_type(self, api_key)-> bool:
+        """
+        Determines if the provided API key and secret are legacy or new.
+        Returns True if legacy, False if new.
+        """
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/coinbase.py:179]: Use self-explanatory function names

```suggestion
    def is_legacy_key(self, api_key)-> bool:
        """Determines if the provided API key is a legacy one"""
```

File: rotkehlchen/exchanges/coinbase.py [hunk 4]
```diff
@@ -160,21 +161,117 @@ def __init__(
...
+        legacy_pattern = r'^[\w]+$' 
+        new_pattern = r'^organizations/[\w-]+/apiKeys/[\w-]+$'
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/coinbase.py:181]: Both of these. `legacy_pattern` and `new_pattern` are constants. Define them at the top of the file and not have them always redefined every time this function is called.

In fact, you can do even better. Since it's a regex pattern you can define and compile it as a constant pattern.

```
LEGACY_RE: Final = re.compile(r'^[\w]+$')
NEW_RE: Final = re.compile(r'^organizations/[\w-]+/apiKeys/[\w-]+$')
```

Move those in the top of the file, and have them defined once. Then use them in the function below like:

`LEGACY_RE.match(api_key)` etc.

File: rotkehlchen/exchanges/coinbase.py [hunk 5]
```diff
@@ -160,21 +161,117 @@ def __init__(
...
+
+        if re.match(legacy_pattern, api_key):
+            log.debug("Legacy Key format!")
+            return True
+        elif re.match(new_pattern, api_key):
+            log.debug("New Key format!")
+            return False
+        else:
+            raise ValueError(f'Invalid API key format: {api_key}')
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/coinbase.py:190]: Our codebase never raises ValueErrors. This would end up being unhandled and crash the application.

Use `AuthenticationError` from `rotkehlchen/errors/api.py`.

Also:

1. Add it in the function's docstring it may raise this.
2. Hande it wherever it's raised.

File: rotkehlchen/exchanges/coinbase.py [hunk 6]
```diff
@@ -160,21 +161,117 @@ def __init__(
             secret=secret,
             database=database,
         )
+        self.is_legacy_api_key=self.determine_key_type(api_key)
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/coinbase.py:164]: Do not call this function here. At initialization of all exchanges `validate_api_key` is always called so any problems with the api key will be caught there. While here the exception will go unhandled.

File: rotkehlchen/exchanges/coinbase.py [hunk 7]
```diff
@@ -160,21 +161,117 @@ def __init__(
...
         self.first_connection_made = True
 
-    def edit_exchange_credentials(self, credentials: ExchangeAuthCredentials) -> bool:
-        changed = super().edit_exchange_credentials(credentials)
-        if credentials.api_key is not None:
-            self.session.headers.update({'CB-ACCESS-KEY': self.api_key})
-        return changed
+    
+
+    # CDP Method
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/coinbase.py:197]: All function related comments should be in the docstring

```suggestion
```

File: rotkehlchen/exchanges/coinbase.py [hunk 8]
```diff
@@ -160,21 +161,117 @@ def __init__(
...
+            raise RemoteError(f'Error generating JWT token: {str(e)}') from e
+        
+    def validate_api_key(self) -> tuple[bool, str]:
+        """
+        Validates that the Coinbase API key is good for usage in rotki.
+        Makes sure that the following permissions are given to the key:
+        wallet:accounts:read, wallet:transactions:read, wallet:withdrawals:read, wallet:deposits:read
+        """
+        if self.is_legacy_api_key:
+            # Validate legacy API key
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/coinbase.py:249]: useless comment

```suggestion
        if self.is_legacy_api_key:
```

File: rotkehlchen/exchanges/coinbase.py [hunk 9]
```diff
@@ -160,21 +161,117 @@ def __init__(
...
+            if len(account_info) != 0:
+                # and now try to get all transactions of an account to see if that's possible
+                method = f'accounts/{account_info[0][0]}/transactions'
+                result, msg = self._validate_single_api_key_action(method)
+                if result is None:
+                    return False, msg
 
+            return True, ''
+        else:
+            # Validate new API key format
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/coinbase.py:265]:
```suggestion
        else:  # Validate new API key format
```

File: rotkehlchen/exchanges/coinbase.py [hunk 10]
```diff
@@ -160,21 +161,117 @@ def __init__(
...
+            api_key_pattern = r'^organizations/[\w-]+/apiKeys/[\w-]+$'
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/coinbase.py:266]: This should already be defined as a constant in the top of the file as a I said above, so use it here and not redefine it

File: rotkehlchen/exchanges/coinbase.py [hunk 11]
```diff
@@ -160,21 +161,117 @@ def __init__(
...
+            if not re.match(api_key_pattern, self.api_key):
+                return False, 'Invalid Coinbase API key name format'
+
+            private_key_pattern = r'^-----BEGIN EC PRIVATE KEY-----\n[\w+/=\n]+-----END EC PRIVATE KEY-----\n?$'
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/coinbase.py:270]: Do the same with this as I asked with other regexes.

File: rotkehlchen/exchanges/coinbase.py [hunk 12]
```diff
@@ -160,21 +161,117 @@ def __init__(
...
+            if not re.match(private_key_pattern, self.secret.decode('utf-8'), re.MULTILINE):
+                return False, 'Invalid Coinbase private key format'
+            return True, ''
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/coinbase.py:273]: readability for last return

```suggestion

            return True, ''
```

File: rotkehlchen/exchanges/coinbase.py [hunk 13]
```diff
@@ -220,27 +317,27 @@ def _validate_single_api_key_action(
...
-        if len(account_info) != 0:
-            # and now try to get all transactions of an account to see if that's possible
-            method = f'accounts/{account_info[0][0]}/transactions'
-            result, msg = self._validate_single_api_key_action(method)
-            if result is None:
-                return False, msg
+    def edit_exchange_credentials(self, credentials: ExchangeAuthCredentials) -> bool:
+        changed = super().edit_exchange_credentials(credentials)
+        if credentials.api_key is not None:
+            new_is_legacy = self.determine_key_type(credentials.api_key)
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/coinbase.py:327]: here you would need to catch the authentication error that may be raised

File: rotkehlchen/exchanges/coinbase.py [hunk 14]
```diff
@@ -220,27 +317,27 @@ def _validate_single_api_key_action(
...
+                self.is_legacy_api_key = new_is_legacy
+                changed = True
+            if self.is_legacy_api_key:
+                self.session.headers.update({'CB-ACCESS-KEY': credentials.api_key})
+            else:
+                self.api_key = credentials.api_key
+        if credentials.api_secret is not None:
+            self.secret = credentials.api_secret
+            changed = True
+        return changed
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/coinbase.py:339]: readability

```suggestion

        return changed
```

File: rotkehlchen/exchanges/coinbase.py [hunk 15]
```diff
@@ -220,27 +317,27 @@ def _validate_single_api_key_action(
...
-            method = f'accounts/{account_info[0][0]}/transactions'
-            result, msg = self._validate_single_api_key_action(method)
-            if result is None:
-                return False, msg
+    def edit_exchange_credentials(self, credentials: ExchangeAuthCredentials) -> bool:
+        changed = super().edit_exchange_credentials(credentials)
+        if credentials.api_key is not None:
+            new_is_legacy = self.determine_key_type(credentials.api_key)
+            if new_is_legacy != self.is_legacy_api_key:
+                # Key type has changed
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/coinbase.py:329]:
```suggestion
            if new_is_legacy != self.is_legacy_api_key:  # Key type has changed
```

File: rotkehlchen/exchanges/coinbase.py [hunk 16]
```diff
@@ -285,34 +382,40 @@ def _api_query(
             options: dict[str, Any] | None = None,
             ignore_pagination: bool = False,
     ) -> list[Any]:
-        """Performs a coinbase API Query for endpoint
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/coinbase.py:288]: Why did you remove the docstring here? If you want to adjust it, do that. If not then don't touch it.

File: rotkehlchen/exchanges/coinbase.py [hunk 17]
```diff
@@ -285,34 +382,40 @@ def _api_query(
...
-
-        You can optionally provide extra arguments to the endpoint via the options argument.
-        If you want just the first results then set ignore_pagination to True.
-        """
         all_items: list[Any] = []
         had_4xx = False
         request_verb = 'GET'
-        # initialize next_uri before loop
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/coinbase.py:296]: Why did you remove this comment?

File: rotkehlchen/exchanges/coinbase.py [hunk 18]
```diff
@@ -285,34 +382,40 @@ def _api_query(
...
+
+                self.session.headers.update({
+                    'CB-ACCESS-SIGN': signature,
+                    'CB-ACCESS-TIMESTAMP': timestamp,
+                })
+            else:
+                request_path = f'{self.host}/{self.apiversion}/{endpoint}'
+                log.debug('CDP Request Path ', request_url=request_path)
+                uri = f'{request_verb} {request_path}'
+                log.debug('CDP URI Path ', request_url=uri)
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/coinbase.py:414]: Don't spam the logs with multiple log entries for the samething.

```suggestion
                uri = f'{request_verb} {self.host}/{self.apiversion}/{endpoint}'
                log.debug('Coinbase CDP API query ', request_url=uri)
```

File: rotkehlchen/exchanges/coinbase.py [hunk 19]
```diff
@@ -325,9 +428,6 @@ def _api_query(
                 gevent.sleep(.5)
                 continue  # do a single retry since they don't have info on retries
 
-            if response.status_code == 403:
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/coinbase.py:328]: Are you absolutely certain that neither the legacy, nor the new API will never ever return a 403 HTTP API status? We had this there for a reason, and that is that theold API at some point did return 403.

File: rotkehlchen/exchanges/coinbase.py [hunk 20]
```diff
@@ -344,22 +444,15 @@ def _api_query(
             if 'data' not in json_ret:
                 raise RemoteError(f'Coinbase json response does not contain data: {response.text}')
 
-            # `data` attr is a list in itself
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/coinbase.py:347]: Here and in all the other lines why are you removing all these comments? What is the purpose of this change?

File: rotkehlchen/exchanges/coinbase.py (hunk 6 above)
Comment (LefterisJP) [inline rotkehlchen/exchanges/coinbase.py:164]:
```suggestion
        self.is_legacy_api_key = self.determine_key_type(api_key)
```

**Review by LefterisJP — COMMENTED** [review]
Ooof ...  lots of comments. And some generic questions.

1. Have you tested that this works?
2. Please run `make lint` before commiting. There is a lot of linting errors
3. Someone from the frontend will check the frontend diff
4. Please push all changes as new commits or rewriting old commits in this PR. Do not open new PRS.

File: rotkehlchen/exchanges/coinbase.py (hunk 2 above)
Comment (vijithreddy) [reply rotkehlchen/exchanges/coinbase.py:176]: We did successfully test the integration with both the keys!
Thanks for your suggestions, made changes.

File: frontend/app/src/components/settings/api-keys/exchange/ExchangeKeysForm.vue [hunk 21]
```diff
@@ -262,7 +271,8 @@ function onExchangeChange(exchange: string) {
       :error-messages="toMessages(v$.apiKey)"
       data-cy="api-key"
       prepend-icon="key-line"
-      :label="t('exchange_settings.inputs.api_key')"
+      :label="exchange.location === 'coinbase' ? 'API Key Name' : t('exchange_settings.inputs.api_key')"
+      :hint="exchange.location === 'coinbase' ? 'Format: organizations/{org_id}/apiKeys/{key_id}' : ''"
```
Comment (lukicenturi) [inline frontend/app/src/components/settings/api-keys/exchange/ExchangeKeysForm.vue:275]: The `API Key Name` need to be put on the locale file `frontend/app/src/locales/en.json`

File: frontend/app/src/components/settings/api-keys/exchange/ExchangeKeysForm.vue [hunk 22]
```diff
@@ -274,9 +284,9 @@ function onExchangeChange(exchange: string) {
       :error-messages="toMessages(v$.apiSecret)"
       data-cy="api-secret"
       prepend-icon="lock-line"
-      :label="t('exchange_settings.inputs.api_secret')"
+      :label="exchange.location === 'coinbase' ? 'Private Key' : t('exchange_settings.inputs.api_secret')"
```
Comment (lukicenturi) [inline frontend/app/src/components/settings/api-keys/exchange/ExchangeKeysForm.vue:287]: same as above

File: frontend/app/src/components/settings/api-keys/exchange/ExchangeKeysForm.vue [hunk 23]
```diff
@@ -289,6 +299,12 @@ function onExchangeChange(exchange: string) {
       :label="t('exchange_settings.inputs.passphrase')"
     />
 
+    <div v-if="exchange.location === 'coinbase'" class="mt-2">
+      <a href="https://portal.cdp.coinbase.com/access/api" target="_blank" rel="noopener noreferrer">
+    {{ t('exchange_keys_form.generate_keys_link') }}
+      </a>
+    </div>
+
```
Comment (lukicenturi) [inline frontend/app/src/components/settings/api-keys/exchange/ExchangeKeysForm.vue:307]: I guess this part is too generic, if we add this one, we need to add it for all exchanges.
So I think just remove it.

I can fix the frontend issues, if you give me access to push to the repo forked in your account.
Or else you need to fix all those issues, and also fix the lint `pnpm run lint:fix`

File: rotkehlchen/exchanges/coinbase.py [hunk 24]
```diff
@@ -160,20 +167,108 @@ def __init__(
...
 
     def first_connection(self) -> None:
         self.first_connection_made = True
 
-    def edit_exchange_credentials(self, credentials: ExchangeAuthCredentials) -> bool:
-        changed = super().edit_exchange_credentials(credentials)
-        if credentials.api_key is not None:
-            self.session.headers.update({'CB-ACCESS-KEY': self.api_key})
-        return changed
+    # CDP Method
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/coinbase.py:197]: As I already mentioned here: https://github.com/rotki/rotki/pull/7903#discussion_r1593533299 removethis comment. You can include it in the docstring itself

```suggestion
```

File: rotkehlchen/exchanges/coinbase.py [hunk 25]
```diff
@@ -160,20 +167,108 @@ def __init__(
             secret=secret,
             database=database,
         )
+
+        self.is_legacy_api_key = self.is_legacy_key(api_key)
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/coinbase.py:171]: As I wrote here: https://github.com/rotki/rotki/pull/7903#discussion_r1593532832 do not call this here.
Read that comment to understand why.
```suggestion

```

File: rotkehlchen/exchanges/coinbase.py [hunk 26]
```diff
@@ -220,27 +315,28 @@ def _validate_single_api_key_action(
...
-    def validate_api_key(self) -> tuple[bool, str]:
-        """Validates that the Coinbase API key is good for usage in rotki
+    def edit_exchange_credentials(self, credentials: ExchangeAuthCredentials) -> bool:
+        changed = super().edit_exchange_credentials(credentials)
+        if credentials.api_key is not None:
+            try:
+                new_is_legacy = self.is_legacy_key(credentials.api_key)
+            except AuthenticationError as e:
+                log.error(f'Invalid API key format: {e}')
+                return False
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/coinbase.py:325]: I think this is the fault of our code/docstring but `changed` here means if a new api key is given for editing. And here it is, so you should return `True`. And then this will fail here: https://github.com/rotki/rotki/blob/252891f9fcf7dbe07e4bdfa7bd261792ae203a90/rotkehlchen/exchanges/manager.py#L100 and let the user know.

File: rotkehlchen/exchanges/coinbase.py [hunk 27]
```diff
@@ -220,27 +315,28 @@ def _validate_single_api_key_action(
...
 
-    def validate_api_key(self) -> tuple[bool, str]:
-        """Validates that the Coinbase API key is good for usage in rotki
+    def edit_exchange_credentials(self, credentials: ExchangeAuthCredentials) -> bool:
+        changed = super().edit_exchange_credentials(credentials)
+        if credentials.api_key is not None:
+            try:
+                new_is_legacy = self.is_legacy_key(credentials.api_key)
+            except AuthenticationError as e:
+                log.error(f'Invalid API key format: {e}')
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/coinbase.py:324]:
```suggestion
                log.error(f'Invalid coinbase API key format: {e}')
```

File: rotkehlchen/exchanges/coinbase.py [hunk 28]
```diff
@@ -220,27 +315,28 @@ def _validate_single_api_key_action(
...
+            if new_is_legacy != self.is_legacy_api_key:  # Key type has changed
+                self.is_legacy_api_key = new_is_legacy
+                changed = True
 
-        return True, ''
+            if self.is_legacy_api_key:
+                self.session.headers.update({'CB-ACCESS-KEY': credentials.api_key})
+            else:
+                self.api_key = credentials.api_key
+        if credentials.api_secret is not None:
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/coinbase.py:335]: readability

```suggestion

        if credentials.api_secret is not None:
```

File: rotkehlchen/exchanges/coinbase.py [hunk 29]
```diff
@@ -285,49 +381,56 @@ def _api_query(
             options: dict[str, Any] | None = None,
             ignore_pagination: bool = False,
     ) -> list[Any]:
-        """Performs a coinbase API Query for endpoint
+        """
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/coinbase.py:384]: Please undo this change as I asked before: https://github.com/rotki/rotki/pull/7903/files#r1593544499

File: rotkehlchen/exchanges/coinbase.py [hunk 30]
```diff
@@ -285,49 +381,56 @@ def _api_query(
...
 
             full_url = self.base_uri + next_uri
+
             try:
                 response = self.session.get(full_url, timeout=timeout)
             except requests.exceptions.RequestException as e:
                 raise RemoteError(f'Coinbase API request failed due to {e!s}') from e
 
-            if response.status_code in (401, 429) and had_4xx is False:
+            if response.status_code in (401, 403, 429) and had_4xx is False:
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/coinbase.py:429]: hmm you resolved this: https://github.com/rotki/rotki/pull/7903#discussion_r1593552039 comment without explaining.

Why did you move it here? What's the reason?

Before if we got a 403 (FORBIDDEN) error we raised a `CoinbasePermissionError`. Meaning that the api key does not have permission for the resource they are trying to access.

Now with your change a (403 FORBIDDEN) just retries. That seems like a wrong change. Can you explain the logic ofthe change here?

File: rotkehlchen/exchanges/coinbase.py [hunk 31]
```diff
@@ -344,22 +447,15 @@ def _api_query(
...
+            all_items.extend(json_ret['data'])  # `data` attr is a list in itself
             if ignore_pagination or 'pagination' not in json_ret:
-                # break out of the loop, no need to handle pagination
-                break
+                break  # break out of the loop, no need to handle pagination
 
             if 'next_uri' not in json_ret['pagination']:
                 raise RemoteError('Coinbase json response contained no "next_uri" key')
 
-            # otherwise, let the loop run to gather subsequent queries
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/coinbase.py:356]: Again as I said before. Why are you removing this comment here? Please do not touch it.

**Review by LefterisJP — COMMENTED** [review]
Thanks for the PR!

Still lots of comments left and especially some which have not been addressed from before.

Also a really serious issue which needs changing is that the PR is now pulling develop into bugfixes. Which is not something you shoiuld do. You need to rebase your commits onto bugfixes, not just change the branch target.

Please make sure to do this while keeping the PR open, otherwise we are going to lose the PR review comments.

File: rotkehlchen/exchanges/coinbase.py (hunk 6 above)
Comment (vijithreddy) [reply rotkehlchen/exchanges/coinbase.py]: I initially removed the comments when I was updating the code to support only the new key format, and I removed all the legacy key code at that time. When the requirement changed to support both key formats, I reinstated the logic for both but inadvertently didn't include the original comments. I believe I should now have added all the relevant comments back into the code.

File: rotkehlchen/exchanges/coinbase.py (hunk 26 above)
Comment (vijithreddy) [reply rotkehlchen/exchanges/coinbase.py]: I believe I made necessary changes!

File: rotkehlchen/exchanges/coinbase.py (hunk 29 above)
Comment (vijithreddy) [reply rotkehlchen/exchanges/coinbase.py:386]: I lost this conversation due to issues with Rebase.

File: rotkehlchen/exchanges/coinbase.py (hunk 30 above)
Comment (vijithreddy) [reply rotkehlchen/exchanges/coinbase.py]: Yeah, makes sense. Fixed the code.

File: rotkehlchen/exchanges/coinbase.py (hunk 31 above)
Comment (vijithreddy) [reply rotkehlchen/exchanges/coinbase.py]: Added back comments! Thank you.

File: rotkehlchen/exchanges/coinbase.py [hunk 32]
```diff
@@ -344,22 +447,15 @@ def _api_query(
...
-            # this next_uri will be used in next iteration
             next_uri = json_ret['pagination']['next_uri']
             if not next_uri:
-                # As per the docs: https://developers.coinbase.com/api/v2?python#pagination
```
[the remark this answers is not in the document]
Comment (vijithreddy) [reply rotkehlchen/exchanges/coinbase.py]: Added back comments! Thank you.

File: rotkehlchen/exchanges/coinbase.py (hunk 25 above)
Comment (vijithreddy) [reply rotkehlchen/exchanges/coinbase.py]: I've incorporated the method call within validate_api_key. However, I'm encountering unexpected behavior during local testing. It appears that validate_api_key does not execute when I restart the app locally with credentials already saved. Editing or re-adding the keys resolves the issue, and logging out and back in also bypasses the problem. However, declaring it at the instance level seems to solve this. Is this behavior expected?

File: rotkehlchen/exchanges/coinbase.py (hunk 24 above)
Comment (vijithreddy) [reply rotkehlchen/exchanges/coinbase.py]: This should be done!

File: rotkehlchen/exchanges/coinbase.py (hunk 29 above)
Comment (vijithreddy) [reply rotkehlchen/exchanges/coinbase.py:386]: Just a heads up on why you might not see the old commits: I attempted to cherry-pick my previous commits, but this process led to corruption in my fork when I tried to rebase. To resolve this and clean up the repository, I ended up losing those old commits.

File: rotkehlchen/exchanges/coinbase.py (hunk 28 above)
Comment (vijithreddy) [reply rotkehlchen/exchanges/coinbase.py:338]: This should be corrected!

File: rotkehlchen/exchanges/coinbase.py (hunk 27 above)
Comment (vijithreddy) [reply rotkehlchen/exchanges/coinbase.py]: Added this suggestion!

File: frontend/app/src/components/settings/api-keys/exchange/ExchangeKeysForm.vue (hunk 21 above)
Comment (vijithreddy) [reply frontend/app/src/components/settings/api-keys/exchange/ExchangeKeysForm.vue]: Added this, thanks for the suggestion!

File: frontend/app/src/components/settings/api-keys/exchange/ExchangeKeysForm.vue (hunk 22 above)
Comment (vijithreddy) [reply frontend/app/src/components/settings/api-keys/exchange/ExchangeKeysForm.vue]: Added this, thanks for the suggestion!

File: frontend/app/src/components/settings/api-keys/exchange/ExchangeKeysForm.vue (hunk 23 above)
Comment (vijithreddy) [reply frontend/app/src/components/settings/api-keys/exchange/ExchangeKeysForm.vue]: Removed this block. I also gave you permissions to edit the repo, please let me know if you are able to access it.

File: rotkehlchen/exchanges/coinbase.py (hunk 25 above)
Comment (LefterisJP) [reply rotkehlchen/exchanges/coinbase.py]: Hmm @vijithreddy  okay then this only runs when the exchange is edited or setup. Then add it here too **but* make sure to add it under try catch for the errors the function adds. And log an error if it happens and set it to legacy.

File: rotkehlchen/exchanges/coinbase.py [hunk 33]
```diff
@@ -160,20 +172,106 @@ def __init__(
             secret=secret,
             database=database,
         )
+
+        self.is_legacy_api_key: bool | None = None
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/coinbase.py:176]: Hmm @vijithreddy okay then this only runs when the exchange is edited or setup. Then add it here too **but make sure to add it under try catch for the errors the function adds**. And log an error if it happens and set it to legacy.

as I wrote here: https://github.com/rotki/rotki/pull/7903#discussion_r1596391668

File: rotkehlchen/exchanges/coinbase.py [hunk 34]
```diff
@@ -220,27 +318,28 @@ def _validate_single_api_key_action(
...
-        account_info = self._get_active_account_info(result)
-        if len(account_info) != 0:
-            # and now try to get all transactions of an account to see if that's possible
-            method = f'accounts/{account_info[0][0]}/transactions'
-            result, msg = self._validate_single_api_key_action(method)
-            if result is None:
-                return False, msg
+            if new_is_legacy != self.is_legacy_api_key:  # Key type has changed
+                self.is_legacy_api_key = new_is_legacy
+                changed = True
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/coinbase.py:331]: no need for this. If you get in here it will always be true due to `credentials.api_key is not None`

File: rotkehlchen/exchanges/coinbase.py (hunk 29 above)
Comment (LefterisJP) [reply rotkehlchen/exchanges/coinbase.py:387]: Hmmm this is not undone. You still are touching this docstring by just adding a newline and nothing else.

File: rotkehlchen/exchanges/coinbase.py [hunk 35]
```diff
@@ -345,7 +456,7 @@ def _api_query(
                 raise RemoteError(f'Coinbase json response does not contain data: {response.text}')
 
             # `data` attr is a list in itself
-            all_items.extend(json_ret['data'])
+            all_items.extend(json_ret['data'])  # `data` attr is a list in itself
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/coinbase.py:459]: If you move `# `data` attr is a list in itself` here then delete it from the line above, please.

**Review by LefterisJP — COMMENTED** [review]
Hey @vijithreddy please check my response to your comment on what you wanted to use in the constructor.

Also tests are failing. And frontend lint is failing.

File: rotkehlchen/exchanges/coinbase.py (hunk 33 above)
Comment (vijithreddy) [reply rotkehlchen/exchanges/coinbase.py]: Added the logic and logging error. 

Also believe I fixed frontend lint issues. And tests failed because of a missing PyJWT module, added that to the requirements.

File: rotkehlchen/exchanges/coinbase.py (hunk 34 above)
Comment (vijithreddy) [reply rotkehlchen/exchanges/coinbase.py:340]: Removed this, thanks for the suggestion!

File: rotkehlchen/exchanges/coinbase.py (hunk 35 above)
Comment (vijithreddy) [reply rotkehlchen/exchanges/coinbase.py:463]: Done! That was an oversight from my end.

File: frontend/app/src/locales/cn.json [hunk 36]
```diff
@@ -2055,8 +2055,8 @@
       "sync_enabled": "同步启用"
     },
     "inputs": {
-      "api_key": "API Key",
-      "api_secret": "API Secret",
+      "private_key": "Private Key",
```
Comment (LefterisJP) [inline frontend/app/src/locales/cn.json:2058]: Why are you changing the kraken account related text? You should not touch these stuff for any locale.

File: rotkehlchen/exchanges/coinbase.py [hunk 37]
```diff
@@ -220,27 +327,24 @@ def _validate_single_api_key_action(
...
-        account_info = self._get_active_account_info(result)
-        if len(account_info) != 0:
-            # and now try to get all transactions of an account to see if that's possible
-            method = f'accounts/{account_info[0][0]}/transactions'
-            result, msg = self._validate_single_api_key_action(method)
-            if result is None:
-                return False, msg
+            if new_is_legacy != self.is_legacy_api_key:  # Key type has changed
+                self.is_legacy_api_key = new_is_legacy
+                changed = True
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/coinbase.py:340]: This is still here @vijithreddy. Please remove as asked here: https://github.com/rotki/rotki/pull/7903/files#r1596587452
```suggestion
```

File: rotkehlchen/exchanges/coinbase.py [hunk 38]
```diff
@@ -285,36 +389,48 @@ def _api_query(
             options: dict[str, Any] | None = None,
             ignore_pagination: bool = False,
     ) -> list[Any]:
-        """Performs a coinbase API Query for endpoint
+        """
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/coinbase.py:392]: Please @vijithreddy this is the 3rd time I ask you.

https://github.com/rotki/rotki/pull/7903/files#r1593544499

 Undo this formatting change. There is no need for it.

**Review by LefterisJP — COMMENTED** [review]
Some more comments @vijithreddy. Most of them are things not done from before. 

As I said we can also take it up and finish it if it's easier for you.

File: rotkehlchen/exchanges/coinbase.py (hunk 38 above)
Comment (vijithreddy) [reply rotkehlchen/exchanges/coinbase.py:392]: I didn't realize I formatted by adding a new line. Apologies, reverting the change

File: frontend/app/src/locales/cn.json (hunk 36 above)
Comment (vijithreddy) [reply frontend/app/src/locales/cn.json:2058]: Apologies, eslint was throwing error. I am reverting this change and adding the changes specific to Coinbase's new Keys

File: rotkehlchen/exchanges/coinbase.py (hunk 38 above)
Comment (vijithreddy) [reply rotkehlchen/exchanges/coinbase.py]: Issued a PR to rever the formatting change

**Review by LefterisJP — COMMENTED** [review]
hey @vijithreddy I am sorry but still the stuff I asked you to do are not done. At this point my team has wasted more time doing reviewing the PR 6 times than it would have taken us to finish this on our own.

We are going to take it over and finish it. Can't be asking you to fix something and then with each new push having reviews ignored and new changes in various files of the app necessitating more reviews and more fixes.

File: frontend/app/src/locales/cn.json (hunk 36 above)
Comment (LefterisJP) [reply frontend/app/src/locales/cn.json:2060]: it's not fixed. Taking over.

File: rotkehlchen/exchanges/coinbase.py [hunk 39]
```diff
@@ -160,20 +172,113 @@ def __init__(
...
+        wallet:accounts:read, wallet:transactions:read,
+        wallet:withdrawals:read, wallet:deposits:read
+
+        For CDP keys, make sure they are formatted properly
+        """
+        self.is_legacy_api_key = self.is_legacy_key(self.api_key)
+        if self.is_legacy_api_key:
             self.session.headers.update({'CB-ACCESS-KEY': self.api_key})
-        return changed
+            self.session.headers.update({'CB-VERSION': CB_VERSION})
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/coinbase.py]: This can be combined with the above. One update is only needed. Doing it.

File: rotkehlchen/exchanges/coinbase.py [hunk 40]
```diff
@@ -160,20 +172,113 @@ def __init__(
...
+        try:
+            self.is_legacy_api_key = self.is_legacy_key(api_key)
+        except AuthenticationError as e:
+            self.is_legacy_api_key = True
+            log.error(f'Error determining API key format: {e}. Defaulting to legacy key.')
+
+        # Set CB-ACCESS-KEY header if its a legacy key format
+        if self.is_legacy_api_key:
+            self.session.headers.update({'CB-ACCESS-KEY': self.api_key})
+            self.session.headers.update({'CB-VERSION': CB_VERSION})
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/coinbase.py]: Shoudl be combined. Doing it

```suggestion
            self.session.headers.update({'CB-ACCESS-KEY': self.api_key, 'CB-VERSION': CB_VERSION})
```

**Review by LefterisJP — APPROVED** [review]
LGTM 


The stuck tests are green but for some reason not reporting correctly as finished.
